lib/scholar/linear/logistic_regression.ex

```diff
@@ -0,0 +1,193 @@
+      )
+
+    fit_verify(x, y, opts)
+    if opts[:num_classes] < 3, do: fit_binary(x, y, opts), else: fit_multinomial(x, y, opts)
+  end
+
+  # Function checks validity of the provided data
+
+  deftransformp fit_verify(x, y, opts) do
+    unless is_integer(opts[:num_classes]) and opts[:num_classes] > 0 do
+      raise ArgumentError,
+            "expected :num_classes to be a positive integer, got: #{inspect(opts[:num_classes])}"
+    end
+
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            "expected x to have shape {n_samples, n_features}, got tensor with shape: #{inspect(Nx.shape(x))}"
+    end
+
+    if Nx.rank(y) != 1 do
+      raise ArgumentError,
+            "expected y to have shape {n_samples}, got tensor with shape: #{inspect(Nx.shape(y))}"
+    end
+
+    unless is_number(opts[:lr]) and opts[:lr] > 0 do
+      raise ArgumentError, "expected :lr to be a positive number, got: #{inspect(opts[:lr])}"
+    end
+
+    unless is_integer(opts[:iterations]) and opts[:iterations] > 0 do
+      raise ArgumentError,
+            "expected :iterations to be a positive integer, got: #{inspect(opts[:iterations])}"
+    end
+  end
+
+  # Binary logistic regression
+
+  defnp fit_binary(x, y, opts \\ []) do
+    iterations = opts[:iterations]
+    lr = opts[:lr]
+    x_t = Nx.transpose(x)
+    y_t = Nx.transpose(y)
+
+    {_m, n} = x.shape
+    coeff = Nx.broadcast(Nx.tensor(0, type: {:f, 32}), {n})
+
+    {_, _, _, _, _, _, final_coeff, final_bias} =
+      while {iter = 0, x, lr, iterations, x_t, y_t, coeff, bias = Nx.tensor(0, type: {:f, 32})},
+            Nx.less(iter, iterations) do
+        {coeff, bias} = update_coefficients(x, x_t, y_t, {coeff, bias}, lr)
+        {iter + 1, x, lr, iterations, x_t, y_t, coeff, bias}
+      end
+
+    %__MODULE__{coefficients: final_coeff, bias: final_bias, mode: 0}
+  end
+
+  # Function computes one-hot encoding
+
+  defnp one_hot_encoding(labels, num_classes) do
+    Nx.equal(Nx.new_axis(labels, -1), Nx.iota({1, num_classes}))
+  end
+
+  # Multinomial logistic regression
+
+  defnp fit_multinomial(x, y, opts) do
+    {_m, n} = x.shape
+    iterations = opts[:iterations]
+    lr = opts[:lr]
+    num_classes = opts[:num_classes]
+    one_hot = one_hot_encoding(y, num_classes)
+    x_t = Nx.transpose(x)
+
+    {_, _, _, _, _, _, _, final_coeff} =
+      while {iter = 0, x, lr, n, iterations, one_hot, x_t,
+             coeff = Nx.broadcast(Nx.tensor(0, type: {:f, 32}), {n, num_classes})},
+            Nx.less(iter, iterations) do
+        coeff = update_coefficients_multinomial(x, x_t, one_hot, coeff, lr)
+        {iter + 1, x, lr, n, iterations, one_hot, x_t, coeff}
+      end
+
+    %__MODULE__{coefficients: final_coeff, bias: Nx.tensor(0, type: {:f, 32}), mode: 1}
+  end
+
+  # Normalized softmax
+
+  defnp softmax(t) do
+    normalized = t - Nx.reduce_max(t, axes: [1], keep_axes: true)
+    Nx.transpose(Nx.transpose(Nx.exp(normalized)) / Nx.sum(Nx.exp(normalized), axes: [1]))
```

Comment:
Let's extract `Nx.exp(normalized)` to a separate variable. I'm not sure if defn compiler would optimise it out, but we would definitely compute it twice when doing the operations individually with an Nx backend.

Reply:
Also, can we transpose/reshape the divisor, instead of transposing twice?

Reply:
I think this will be ok
```elixir 
 defnp softmax(t) do
    normalized_exp = t - Nx.reduce_max(t, axes: [0], keep_axes: true) |> Nx.exp()
    normalized_exp / Nx.sum(normalized_exp, axes: [0])
  end
```